### PR TITLE
Level4 electron

### PR DESCRIPTION
**Pull request by wufe**
I tried to implement the frontend (renderer) process communication with the backend (main) process via commands and events, since with electron 12+ (released in march 2021) the `nodeIntegration` is set to `false` and `contextIsolation` is set to `true` by default for security reasons.
Furthermore the `remote` module is deprecated, because it is known to be a huge performance bottleneck.
So I decided to use `ipcMain` and `ipcRenderer` to exchange events back and forth from the renderer process to the main process, via an exposed secure API available on the frontend side (implemented in `/src/preload/index.ts`).

***

Speaking of the main application logic itself, the TCP file transfer required some sort of format or protocol to exchange files metadata (i.e. *the filename*) in the first place.  

There were multiple choices that could've been used:
- a simple protocol that would've encapsulated the entire message, containing the filename and the binary data in their field
- binary data with a N bytes prefix, where N is a fixed size, containing the filename that would have been wrapped on the right by a padding of 0x00 bytes
- binary data with a variable-size prefix -containing the filename-, where the end of the prefix would have been limited by a special sequence of bytes
- binary data with a variable-size prefix -containing the filename-, where the length of the prefix would have been prepended to the prefix itself (somewhat like UTF8 does)

All these solutions, even though are based on a single message sent, required to
- receive the data-chunks (which are of variable size)
- concatenate their UTF8 string representation
- split the fixed-size prefix (or the variable-size one) from the binary data
- recover the binary data which was converted to string
- change the control flow to save the new chunk data as binary data into its file

*e.g.*:
```
message: <prefix-with-filename(utf8)><binary-data>

chunk1:  
<prefix-with-fi...>  

chunk2:
<...lename(utf8)><bina...>  

chunk3:
<...ry-data>

# In this example the first message would have been parsed as string, the second one as string too (but containing binary data too), the third as binary
```

For these reasons,  I chose to use a commonly agreed protocol instead, inspired by the `multipart/form-data` boundary.
It is cleanly programmable with a **scalable** state machine, that for the sake of simplicity of this repository I decided to omit.

This is the protocol:  

```javascript
/**
 * Metadata exchange protocol:
 * 
 * Legend:
 * - boundary: random identifier
 * - filename: metadata
 * 
 * Server                   Client
 * ----------------------------------------
 *                          <- CONNECT
 * (1) send boundary ->
 *                          -> (2) receive boundary
 *                          <- (3) SEND FILENAME
 *                          <- (4) send boundary    <=====   Can send other metadata here
 * (5) RECEIVE FILENAME <-
 * (6) receive boundary <-
 * (7) send boundary ->
 *                          -> (8) receive boundary
 *                          <- (9) SEND FILE (bin)
 * (10) RECEIVE FILE (bin) <-
 */
```

Basically the server creates, upon client connection, a random identifier called **boundary** and forwards it to the client.  
The client receives this boundary and **uses it as final delimiter** to send the filename (and maybe other metadata).  
Until now, all the message chunks have been parsed as a string.  
Once the server receives the metadata and the delimiter back, it can now be sure that every further data chunk would be the file itself, i.e. binary data.  
So it can now securely parse data as binary data (`Buffer`) and pipe it into a `fs.WriteStream`, delegating to Node.JS the handling of the data.

Since a protocol like this is scalable (can easily allow adding more metadata, or incorporate metadata formatted with JSON), and there are no real performance drawbacks, I prioritized on scalability and clean flow/code.

### CODE REVIEW

File: src/preload/api/server_api.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        });
+    },
+
+    send_file_command(file: string, address: string, port: number): Promise<void>
+    {
+        return new Promise<void>((resolve, reject) =>
+        {
+            ipcRenderer.send(send_file_command, file, address, port);
+
+            function on_send_failed(_: IpcRendererEvent, reason: Error): void
```
Comment (alessiopcc) [inline src/preload/api/server_api.ts]: We prefer arrow functions 🧑‍🍳

File: src/App.ts [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+                {
+                    // (10) Receive the binary data
+                    write_stream.write(data);
+                }
+            });
+        });
+
+        server.on('error', err =>
+        {
+            throw err;
```
Comment (alessiopcc) [inline src/App.ts]: Throwing inside a callback may lead to unexpected behavior 🧑‍🍳

**Review by alessiopcc — COMMENTED** [review]
Very good job!
I really like the implementation. Just two minor comments

**Review by marmos91 — APPROVED** [review]
Very nice!

File: src/preload/api/server_api.ts (hunk 1 above)
Comment (wufe) [reply src/preload/api/server_api.ts]: Reproduced hoisting manually with arrow functions

File: src/App.ts (hunk 2 above)
Comment (wufe) [reply src/App.ts]: Handled exception dispatching visible error to the client side

**Pull request closed without merging**